compiler/src/iree/compiler/Dialect/HAL/IR/HALDialect.cpp

```diff
@@ -115,6 +117,30 @@ class HALToVMConversionInterface : public VMConversionDialectInterface {
   }
 };
 
+class HALAffinityAnalysisDialectInterface
+    : public IREE::Stream::AffinityAnalysisDialectInterface {
+public:
+  using AffinityAnalysisDialectInterface::AffinityAnalysisDialectInterface;
+  IREE::Stream::LayoutAttrSolverFn
+  makeLayoutAttrSolver(ModuleOp moduleOp) const {
+    return [=](IREE::Stream::AffinityAttr aff, Operation *op,
+               SetVector<Attribute> &layoutAttrs) {
+      // TODO: This needs to be in the lambda. Otherwise, it could crash because
+      // the root op (i.e., the original moduleOp) could be outdated.
+      IREE::HAL::DeviceAnalysis deviceAnalysis(moduleOp);
+      if (failed(deviceAnalysis.run())) {
+        op->emitError("failed to run DeviceAnalysis");
+        return failure();
```

Comment:
emitError returns failure
```suggestion
        return op->emitError("failed to run DeviceAnalysis");
```